File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.cc
#include <bits/stdc++.h>
#include <string>
using namespace std;

vector<int> graph[1001];
void dfs(int v) {
	bool visited[1001]{ 0 };
	stack<int> s;
	int startNode = v;
	visited[v] = true;
	cout << v<<" ";
	for (int i = 1; i <= 1000; ++i) {
		sort(graph[i].begin(), graph[i].end(), greater<>());
	}
	for (auto nodeV : graph[v]) {
		s.push(nodeV);
	}
	while (!s.empty()) {
		int next = s.top();
		s.pop();
		if (visited[next])
			continue;
		cout << next << " ";
		visited[next] = true;
		for (auto nodeNext : graph[next]) {
			s.push(nodeNext);
		}
	}
	cout << '\n';
}

void bfs(int v) {
	int visited[1001]{ 0 };
	queue<int> q;
	int startNode = v;
	visited[v] = true;
	cout << v << " ";
	for (int i = 1; i <= 1000; ++i) {
		sort(graph[i].begin(), graph[i].end());
	}
	for (auto nodeV : graph[v]) {
		q.push(nodeV);
	}
	while (!q.empty()) {
		int next = q.front();
		q.pop();
		if (visited[next])
			continue;
		cout << next << " ";
		visited[next] = true;
		for (auto nodeNext : graph[next]) {
			q.push(nodeNext);
		}
	}
	cout << '\n';

}

int main() {
	int n, m,v;
	cin >> n >> m >> v;
	for (int i = 0; i < m; ++i) {
		int node1, node2;
		cin >> node1 >> node2;
		graph[node1].push_back(node2);
		graph[node2].push_back(node1);
	}
	
	dfs(v);
	bfs(v);

}

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys


input = sys.stdin.readline
n,m,v = map(int,input().split())

graph = [[0]* (n+1) for i in range(n+1)]
visited = [0] * (n+1)

for i in range(m):
    a,b = map(int,input().split())
    graph[a][b] = graph[b][a] = 1

def dfs(v):
    stack = [v]
    # dfs 끝나고 나면 다 1되어있을 거니까 0으로 방문 판정
    visited[v] = 0
    while stack:
        v = stack.pop()
        if(visited[v] == 1):
            continue
        visited[v] = 1
        print(v, end = ' ')
        for i in range(n,0,-1):
            if(visited[i] == 0 and graph[v][i] == 1):
                stack.append(i)
                
def bfs(v):
    queue = [v]
    # dfs 끝나고 나면 다 1되어있을 거니까 0으로 방문 판정
    visited[v] = 0
    while queue:
        v = queue.pop(0)
        print(v, end = ' ')
        for i in range(1,n+1):
            if(visited[i] == 1 and graph[v][i] == 1):
                queue.append(i)
                visited[i] = 0

dfs(v)
print()
bfs(v)